scripts/ai/PathfinderController.lua

```diff
@@ -0,0 +1,364 @@
+				self:callCallback(self.callbackRetryFunction, self.lastContext, self.failCount == self.numRetries, self.failCount)
+				return
+			end
+		end
+	end
+	self:callCallback(self.callbackSuccessFunction, retValue == self.SUCCESS_FOUND_VALID_PATH, path, goalNodeInvalid)
+	self:reset()
+end
+
+--- Is the path found and valid?
+---@param path table|nil
+---@param goalNodeInvalid boolean|nil
+---@return integer
+function PathfinderController:isValidPath(path, goalNodeInvalid)
+	if path and #path > 2 then
+        self:debug('Found a path (%d waypoints, after %d ms)', #path, self.timeTakenMs)
+        return self.SUCCESS_FOUND_VALID_PATH
+    end
+	if goalNodeInvalid then
+		self:error('No path found, goal node is invalid')
+		return self.ERROR_INVALID_GOAL_NODE
+	end 
+	self:error("No path found after %d ms", self.timeTakenMs)
+	return self.ERROR_NO_PATH_FOUND
+end
+
+--- Finds a path to given goal node
+---@param context PathfinderControllerContext
+---@param goalNode number
+---@param xOffset number
+---@param zOffset number
+---@return boolean Was path finding started?
+function PathfinderController:findPathToNode(context, 
+	goalNode, xOffset, zOffset)
+
+	if not self.callbackSuccessFunction then
+		self:error("No valid success callback was given!") 
+		return false
+	end
+	self:started(context)
+	local pathfinder, done, path, goalNodeInvalid = PathfinderUtil.startPathfindingFromVehicleToNode(
+		context.vehicle,
+		goalNode,
+		xOffset,
+		zOffset,
+		context.allowReverse, 
+		context.useFieldNum and 1,
+		context.vehiclesToIgnore,
+		context.maxFruitPercent,
+		context.offFieldPenalty,
+		context.areaToAvoid,
+		context.mustBeAccurate
+	)
+	if done then 
+		self:finished(path, goalNodeInvalid)
+	else 
+		self:debug("Continuing as coroutine...")
+		self.pathfinder = pathfinder
+	end
+	return true
+end
+
+--- Finds a path to a waypoint of a course.
+---@param context PathfinderControllerContext
+---@param course Course
+---@param waypointIndex number
+---@param xOffset number
+---@param zOffset number
+---@return boolean Was path finding started?
+function PathfinderController:findPathToWaypoint(context, 
+	course, waypointIndex, xOffset, zOffset)
+
+	if not self.callbackSuccessFunction then 
+		self:error("No valid success callback was given!") 
+		return false
+	end
+	self:started(context)
+	local pathfinder, done, path, goalNodeInvalid = PathfinderUtil.startPathfindingFromVehicleToWaypoint(
+		context.vehicle, 
+		course, 
+		waypointIndex,
+		xOffset,
+		zOffset,
+		context.allowReverse,
+		context.useFieldNum and CpFieldUtil.getFieldNumUnderVehicle(context.vehicle),
+		context.vehiclesToIgnore, 
+		context.maxFruitPercent,
+		context.offFieldPenalty, 
+		context.areaToAvoid, 
+		context.areaToIgnoreFruit)
+	if done then 
+		self:finished(path, goalNodeInvalid)
+	else 
+		self:debug("Continuing as coroutine...")
+		self.pathfinder = pathfinder
+	end
+end
+
+function PathfinderController:getTemporaryCourseFromPath(path)
```

Comment:
why do we have to expose this, can't we just pass a Course in to the success callback?